core/src/test/java/io/confluent/rest/RateLimitNetworkTrafficListenerTest.java

```diff
@@ -0,0 +1,217 @@
+
+import java.net.URI;
+import java.time.Duration;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.core.Configurable;
+import javax.ws.rs.core.Form;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.Status;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInfo;
+
+public class RateLimitNetworkTrafficListenerTest {
+
+  private static TestRestConfig testConfig;
+  private static ApplicationServer<TestRestConfig> server;
+  private static final String TEST_MESSAGE = "Test message";
+
+  private ScheduledExecutorService executor;
+  private TestApp app;
+  private Client client;
+
+  @BeforeEach
+  public void setup(TestInfo info) throws Exception {
+    Properties props = new Properties();
+    props.setProperty(RestConfig.LISTENERS_CONFIG, "http://0.0.0.0:0");
+
+    if (info.getDisplayName().contains("NetworkTrafficRateLimitEnabled")) {
+      props.put(RestConfig.NETWORK_TRAFFIC_RATE_LIMIT_ENABLE_CONFIG, "true");
+      props.put(RestConfig.NETWORK_TRAFFIC_RATE_LIMIT_BYTES_PER_SEC_CONFIG, 90000);
+      if (info.getDisplayName().contains("Resilience4j")) {
+        props.put(RestConfig.NETWORK_TRAFFIC_RATE_LIMIT_BACKEND_CONFIG, "Resilience4j");
+      }
+    }
+
+    testConfig = new TestRestConfig(props);
+    server = new ApplicationServer<>(testConfig);
+    app = new TestApp("/app");
+    server.registerApplication(app);
+    server.start();
+
+    executor = Executors.newScheduledThreadPool(4);
+    client = ClientBuilder.newClient();
+  }
+
+  @AfterEach
+  public void tearDown() throws Exception {
+    server.stop();
+    server.join();
+
+    client.close();
+    awaitTerminationAfterShutdown(executor);
+  }
+
+  @Test
+  @DisplayName("NetworkTrafficRateLimitDisabled")
+  public void testNetworkTrafficRateLimitDisabled_unlimited() throws Exception {
+    long startTime = System.nanoTime();
+    // send 1000 POST requests in 1 second
+    hammerAtConstantRate(app.getServer().getURI(), "/resource", Duration.ofMillis(1), 10, 1000);
+    double durationMillis = (System.nanoTime() - startTime) / 1_000_000.0;
+    // with no rate limit, 1000 requests should finish less than 2 seconds
+    assertThat("Duration must be greater than 1 second", durationMillis >= 1000);
```

Comment:
So my understanding is this.
1. Rate-limit disabled,  1 sec <= time <= 2 sec
2. Rate-limit enabled, time > 2 sec.

Is it possible to increase the time gap between 1 & 2.? Example if rate-limit is enabled, slow down is more drastic > 10 sec. That way test-cases will give high confidence that rate-limit logic is working as expected.

Reply:
@msn-tldr I can do that but that means the CI will take longer 😄

Reply:
Is somewhere around 5seconds good enough to show the effect?

Reply:
> Is somewhere around 5seconds good enough to show the effect?
Works. 

> I can do that easily (by changing the configuration value) but that means the CI will take longer 😄
Its OK if test takes upto 1 minute, as long signal is high confidence.

Reply:
done 👍